txstratum/protocol.py

```diff
@@ -106,18 +119,45 @@ def __init__(self, manager: 'TxMiningManager'):
             'mining.extranonce.subscribe': self.method_extranonce_subscribe,
         }
 
+        self.messages_in_transit: Dict[int, MessageInTransit] = {}
+
     @property
     def miner_type(self) -> str:
         """Return the type of the connected miner."""
         return self.miner_version.split('/')[0]
 
     def handle_result(self, result: Any, msgid: JSONRPCId) -> None:
         """Handle a result from JSONRPC."""
-        if msgid == 'client.get_version':
+        if msgid not in self.messages_in_transit:
+            self.log.warning('Received result for unknown message', msgid=msgid, result=result)
+            return
+
+        assert type(msgid) is int
+
+        message = self.messages_in_transit.pop(cast(int, msgid))
+
+        if message.method == 'client.get_version':
             self.log.info('Miner version: {}'.format(result), miner_id=self.miner_id)
             self.miner_version = result
         else:
-            self.log.error('Cant handle result: {}'.format(result), miner_id=self.miner_id)
+            self.log.error('Cant handle result: {}'.format(result), miner_id=self.miner_id, msgid=msgid)
+
+    def send_and_track_request(self, method: str, params: Any) -> None:
+        """Send a request to the client and track it."""
+        assert len(self.messages_in_transit) < 1000000
+
+        msgid = random.randint(1, 1000000)
+
+        # We need to make sure msg ids are unique
+        while msgid in self.messages_in_transit:
+            msgid = random.randint(1, 1000000)
```

Comment:
I don't think you need a random it, a sequential id works fine, like how we do on the merged-mining coordinator: https://github.com/HathorNetwork/hathor-core/blob/master/hathor/merged_mining/coordinator.py#L342-L347

But either, maybe just making this into a "next_msgid" method would be good enough.

Reply:
Done